### DISCUSSION TITLE
Refresh grid after serenity save event

### DISCUSSION THREAD

**Discussion by wirble** [serenity-is/Serenity#6684] [Help]
Started off this code but need final missing piece #6683 

Objected was to display error thrown to users in the save event.  The below function works but it doesn't refresh the grid after the save event.  Does anyone know what I can add to refresh the grid after wards.  I even tried creating a custom dialog in the grid instead of using [EditLink] detail in this issue (https://github.com/serenity-is/Serenity/issues/3179) but still not good enough.

This kind of override the Super.save().  What is in the default save that refresh the grid?

any suggestions?

           protected getToolbarButtons() {
           let buttons = super.getToolbarButtons();
           buttons.splice(Q.indexOf(buttons, x => x.cssClass == "save-and-close-button"), 1);

            buttons.push({
                title: Q.text('Controls.EntityDialog.SaveButton'),
                cssClass: 'save-and-close-button',
                hotkey: 'alt+s',
                //icon: '',
                onClick: () => {
                    var self = this;
                    if (this.validateForm() && this.validateBeforeSave()) {
                        var options = this.getSaveOptions((response) => {

                        });

                        options.onSuccess = function (response) {
                            self.dialogClose();
                            //self.entityId = response.EntityId;
                            //self.reloadById();
                        }
                        options.onError = function (response) {
                            Q.notifyInfo(response.Error.Message);

                        }

                        Q.serviceCall(options); //Is this neccessary?
                    }
                },
                
                visible: () => !this.isDeleted() && !this.isViewMode(),
                disabled: () => !this.hasSavePermission() || this.readOnly
               
            });
            return buttons;
        }

**Comment by LSNicholls** [comment]
OK I am probably doing this wrong, but I am passing a reference to the grid to the dialog, so the dialog has this:

`protected caller: MyGrid; `

 Then doing this from the dialog:

```
 protected onSaveSuccess(response: SaveResponse): void {
        if (this.caller) {
            this.caller.refresh();
        }
    }
```

I think there is another answer, because the basic samples (DefaultValuesInNewGrid) have code that looks like this:

```
                // let grid watch for changes to manually created dialog, 
                // so when a new item is saved, grid can refresh itself
                this.initDialog(dlg);
```
... and the comments look like that would work the way you want.  I haven't had a chance to test whether it would work for me if I switched to that.

HTH.

**Reply by wirble** [reply]
@LSNicholls thanks for trying but those didn't really help.  I noticed you seemed to have passed a reference of the grid to dialog, I never learned how people did that besides the serenity.extra repo.

Ultimately, I think I was just barking on the wrong tree in what I wanted to do.  I already had a custom service that checks for rules before saved but my lack of skills in typescript didn't know how to call the service and have it wait in the validateBeforeSave event.  i knew I had something to do with async but didn't know how to put it in the option.  But discussion here gave me the final answer needed #6169 

this allow checks to be run, wait for checks, display error message

        protected validateBeforeSave() {
            super.validateBeforeSave();
            var validated = false;
            var myrow = this.entity;
            myrow.AppointmentId = this.form.AppointmentId.selectedItem.Id;
            //only check with backend when pass client side validation.
            if (super.validateBeforeSave()) {
                MyCustomRuleCheckerService.RunCheck({
                    MyClassRow: myrow
                }, response => {
                    
                    validated = !response.IsError;
                    if (response.IsError)
                        Q.notifyError(response.Descriptions);
                    return validated;
                }, {async:false});
            }
            return validated;

        }

**Comment by LSNicholls** [comment]
> I noticed you seemed to have passed a reference of the grid to dialog, I never learned you people did that besides the serenity.extra repo.

I don't know who "you people" are, I'm a novice :-).  And I promise you I know less about TypeScript than almost anybody here... 

I have just started using validateBeforeSave(), for the first time, in a completely different scenario, so thank you for that useful pointer!  I am still not very clear on custom services, how injection works in this framework, etc.  When I write something like a custom editor I get it wrong six different ways before I get it right.

But, holding onto a grid definitely working for me, and it's useful, so here is how I am doing it, and why:

1. I have a custom Clone button on the grid row.  When it is used, I didn't see the grid refreshing the way I wanted it to, when the dialog saved. (Maybe I'm calling initDialog at the wrong time or something, but it didn't help.) That's the "Why".

2. the grid row button code looks like this, in the onClick method, following the examples for how this method should work for custom  buttons in the grid row:
````
 else if (target.hasClass('clone-details')) {
                var dlg = new MyEntityDialog();
                this.initDialog(dlg);
                dlg.loadNewAndOpenDialog();
                dlg.Clone(item, this);
            }
````
3. As you can see, the button is what passes the reference to the grid as part of calling the dialog's public Clone method, which looks like this:
````
    public Clone(modelRow: MyEntityRow, grid: MyEntityGrid) {
        var clone: MyEntityRow = new Object() as any;
        this.caller = grid;
        this.dialogTitle = AddCloneSuffix(this.dialogTitle);
        clone = extend(clone, modelRow);
        this.adjustFieldsForClone(clone);
        clone.Id = null;
        this.loadEntity(clone);
    }
````
4. As a bonus, you see the  "adjustFieldsForClone" call in the Dialog.Clone method.  This is also called when the native Clone button is invoked within the dialog, keeping everything consistent.  

IOW, where the examples have some explicit assignments in the getCloningEntity method, I have this instead:

````
 protected getCloningEntity() {
        var clone = super.getCloningEntity();
        this.adjustFieldsForClone(clone);
        return clone;
    }
````
6. Finally the onSaveSuccess() does what I indicated before and the grid is refreshed properly on save. In my case, there is also a detail grid on the calling form, and it was actually the detail grid that wasn't showing the new contents I expected to see, which might have been the problem with this.initDialog() not doing what I wanted. But in any case they both do now.

hth

**Reply by wirble** [reply]
that was supposed to be "how" not "you", typo...thanks for the additional information.

**Reply by LSNicholls** [reply]
Oh! <rofl>.  In any case, I don't know how other people do it either.  :-)

**Comment by edwardch** [comment]
I saw the title and haven't read all the details, so apologies if I'm on the wrong tangent here. 

I call `SubDialogHelper.triggerDataChange(this);` in my dialog which triggers the grid refresh.

Hope I have the context right :)

**Comment by LSNicholls** [comment]
Wow.  Thank you Edward.  I didn't know that was there.   I clearly have to spend more time with the source code.

I have been trying to figure out how to bind to a change in a drop down, completely different scenario, and this might work for that purpose, too.
